From 786769c42d4af57a172650ba0fa624138e109bce Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 15 Jan 2024 17:38:00 +0100
Subject: [PATCH 0577/1695] plat-stm32mp2: add function do_reset()

Add the new helper function do_reset() to handle the system reset.

This patch is a preliminary step for reset support in TAMPER driver
for STM32MP25.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 220cb13269f6295f0ccf883bdb1c7fa4a1580875 plat-stm32mp2: add function do_reset()]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ie6ad038c073f7af106544e008d5548b547b7b450
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/351403
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c            | 40 +++++++++++++++++++
 core/arch/arm/plat-stm32mp2/platform_config.h |  3 ++
 core/arch/arm/plat-stm32mp2/stm32_util.h      |  3 ++
 3 files changed, 46 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index a93204da6..6ce096ed3 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -6,6 +6,7 @@
 #include <config.h>
 #include <console.h>
 #include <drivers/gic.h>
+#include <drivers/rstctrl.h>
 #include <drivers/stm32_bsec.h>
 #include <drivers/stm32_rif.h>
 #include <drivers/stm32_risab.h>
@@ -16,6 +17,7 @@
 #include <initcall.h>
 #include <kernel/abort.h>
 #include <kernel/boot.h>
+#include <kernel/delay.h>
 #include <kernel/dt.h>
 #include <kernel/interrupt.h>
 #include <kernel/misc.h>
@@ -216,3 +218,41 @@ static TEE_Result init_debug(void)
 }
 early_init_late(init_debug);
 #endif /* CFG_STM32_BSEC3 */
+
+static bool stm32mp_supports_second_core(void)
+{
+	if (CFG_TEE_CORE_NB_CORE == 1)
+		return false;
+
+	return true;
+}
+
+void __noreturn do_reset(const char *str __maybe_unused)
+{
+	struct rstctrl *rstctrl = NULL;
+
+	if (stm32mp_supports_second_core()) {
+		uint32_t target_mask = 0;
+
+		if (get_core_pos() == 0)
+			target_mask = TARGET_CPU1_GIC_MASK;
+		else
+			target_mask = TARGET_CPU0_GIC_MASK;
+
+		interrupt_raise_sgi(interrupt_get_main_chip(),
+				    GIC_SEC_SGI_1, target_mask);
+		/* Wait that other core is halted */
+		mdelay(1);
+	}
+
+	IMSG("Forced system reset %s", str);
+	console_flush();
+
+	/* Request system reset to RCC driver */
+	rstctrl = stm32mp_rcc_reset_id_to_rstctrl(SYS_R);
+	rstctrl_assert(rstctrl);
+	udelay(100);
+
+	/* Can't occur */
+	panic();
+}
diff --git a/core/arch/arm/plat-stm32mp2/platform_config.h b/core/arch/arm/plat-stm32mp2/platform_config.h
index c59e059af..af7ac9c78 100644
--- a/core/arch/arm/plat-stm32mp2/platform_config.h
+++ b/core/arch/arm/plat-stm32mp2/platform_config.h
@@ -110,6 +110,9 @@
 #define TARGET_CPU1_GIC_MASK		BIT(1)
 #define TARGET_CPUS_GIC_MASK		GENMASK_32(CFG_TEE_CORE_NB_CORE - 1, 0)
 
+/* SGI list */
+#define GIC_SEC_SGI_1			9
+
 /* USART/UART resources */
 #define USART1_BASE			UART1_BASE
 #define USART2_BASE			UART2_BASE
diff --git a/core/arch/arm/plat-stm32mp2/stm32_util.h b/core/arch/arm/plat-stm32mp2/stm32_util.h
index aa0e5b464..54a89f75e 100644
--- a/core/arch/arm/plat-stm32mp2/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp2/stm32_util.h
@@ -23,4 +23,7 @@ static inline void stm32mp_register_gpioz_pin_count(size_t count __unused) { }
 #define may_spin_lock(lock)		  cpu_spin_lock_xsave(lock)
 #define may_spin_unlock(lock, exceptions) cpu_spin_unlock_xrestore(lock, \
 								   exceptions)
+
+void __noreturn do_reset(const char *str);
+
 #endif /*__STM32_UTIL_H__*/
-- 
2.47.2

